client-runtime/src/main/java/com/microsoft/rest/v2/http/HttpResponse.java

```diff
@@ -6,43 +6,75 @@
 
 package com.microsoft.rest.v2.http;
 
+import rx.Single;
+
 import java.io.IOException;
 import java.io.InputStream;
 
 /**
  * This class contains all of the details necessary for reacting to a HTTP response from a
  * HttpResponse.
  */
-public interface HttpResponse {
+public abstract class HttpResponse {
     /**
      * Get whether or not this response object has a body.
      * @return Whether or not this response object has a body.
      */
-    boolean hasBody();
+    public abstract boolean hasBody();
 
     /**
      * Get this response object's body as an InputStream. If this response object doesn't have a
      * body, then null will be returned.
      * @return This response object's body as an InputStream. If this response object doesn't have a
      * body, then null will be returned.
      */
-    InputStream bodyAsInputStream();
+    public InputStream bodyAsInputStream() {
+        return bodyAsInputStreamAsync().toBlocking().value();
+    }
+
+    /**
+     * Get this response object's body as an InputStream. If this response object doesn't have a
+     * body, then null will be returned.
+     * @return This response object's body as an InputStream. If this response object doesn't have a
+     * body, then null will be returned.
+     */
+    public abstract Single<? extends InputStream> bodyAsInputStreamAsync();
+
+    /**
+     * Get this response object's body as a byte[]. If this response object doesn't have a body,
+     * then null will be returned.
+     * @return This response object's body as a byte[]. If this response object doesn't have a body,
+     * then null will be returned.
+     * @throws IOException On network error.
+     */
+    public byte[] bodyAsByteArray() throws IOException {
+        return bodyAsByteArrayAsync().toBlocking().value();
+    }
 
     /**
      * Get this response object's body as a byte[]. If this response object doesn't have a body,
      * then null will be returned.
      * @return This response object's body as a byte[]. If this response object doesn't have a body,
      * then null will be returned.
-     * @throws IOException On network issues.
      */
-    byte[] bodyAsByteArray() throws IOException;
+    public abstract Single<byte[]> bodyAsByteArrayAsync();
```

Comment:
It seems to me that if was consuming this interface, then I would expect bodyAsByteArrayAsync() to return the entire response's body as a single byte array. If we want to provided the "user" (the user of these interfaces is only our proxy implementations of our "inner" interfaces, right?) with the option of getting the byte array as chunks, such as bodyAsSegmentedByteArrayAsync() or bodyAsChunkedByteArrayAsync().